### PR TITLE
Latest urls

### PR DESCRIPTION
**Pull request by jaydeland**
Head to open a new PR

### CODE REVIEW

**Comment by jaydeland** [comment]
Still need to introduce a waiting step

**Comment by jaydeland** [comment]
Added TE links:
http://latest.mattermost.com/mattermost-team-linux
http://latest.mattermost.com/mattermost-team-osx
http://latest.mattermost.com/mattermost-team-windows

**Comment by jaydeland** [comment]
Crap - there is no config available in the tests on CircleCI

**Comment by jaydeland** [comment]
Updated with a set of 3 unit tests to ensure routing rules are added as expected

File: server/latest.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+		LogError("Unable to get the %s AWS Bucket Website Config.", cfg.S3BucketNameForLatestURLs)
+		return err
+	}
+
+	// Check for Routing Rules that are not related to the typeToRelease value and carry them forward
+	for _, value := range bucketConfig.RoutingRules {
+		checked := false
+
+		if typeToRelease == "desktop" {
+			checked = (strings.Contains(*value.Condition.KeyPrefixEquals, "enterprise") || strings.Contains(*value.Condition.KeyPrefixEquals, "team"))
```
Comment (metanerd) [inline server/latest.go]:
```suggestion
			checked = strings.Contains(*value.Condition.KeyPrefixEquals, "enterprise") || strings.Contains(*value.Condition.KeyPrefixEquals, "team")
```

File: server/latest.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+			if *result.KeyCount != int64(0) {
+				return result, nil
+			}
+			timeLeft, _ := ctx.Deadline()
+			LogInfo("Release %s is not available; waiting until: %s", releaseBucket+"/"+s3Prefix, timeLeft.Format("15:04:05"))
+		}
+	}
+}
+
+func preserverExistingRoutingRules(svc *s3.S3, cfg *MatterbuildConfig, typeToRelease string, params s3.PutBucketWebsiteInput) error {
```
Comment (metanerd) [inline server/latest.go:142]: typo

File: server/latest.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/pkg/errors"
+)
+
+// Get the existing wensite config if any exists
```
Comment (metanerd) [inline server/latest.go]: typo

File: server/latest.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	return result, nil
+}
+
+func checkBucket(svc *s3.S3, input *s3.ListObjectsV2Input) (*s3.ListObjectsV2Output, error) {
+	result, err := svc.ListObjectsV2(input)
+	return result, err
+}
+
+func checkIfBucketExistsWithPrefixAndWait(ctx context.Context, svc *s3.S3, cfg *MatterbuildConfig, ver string, typeToRelease string) (*s3.ListObjectsV2Output, error) {
+
```
Comment (metanerd) [inline server/latest.go]: Can we remove those empty lines after functions or is it more readable that way? It seems inconsistent with the similar code in this project.

File: server/latest.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+			ReplaceKeyWith: aws.String(file),
+		},
+	}
+	LogInfo("Adding %s RoutingRule for: %s", keyToUse+suffix, file)
+	params.WebsiteConfiguration.RoutingRules = append(params.WebsiteConfiguration.RoutingRules, valueToAdd)
+
+	return nil
+}
+
+func generateNewRoutesForRelease(cfg *MatterbuildConfig, result *s3.ListObjectsV2Output, fileSearchValue string, ver string, params s3.PutBucketWebsiteInput) error {
```
Comment (metanerd) [inline server/latest.go]:
```suggestion
func generateNewRoutesForRelease(result *s3.ListObjectsV2Output, fileSearchValue string, ver string, params s3.PutBucketWebsiteInput) error {
```
It shows up as unused.

File: server/latest.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+	for _, value := range result.Contents {
+		if strings.Contains(*value.Key, fileSearchValue) && !strings.Contains(*value.Key, ".sig") {
+			switchValue := *value.Key
+			switch {
+			case strings.HasSuffix(switchValue, ".dmg"):
+				addRoutingRule(*value.Key, fileSearchValue, params, "-dmg")
```
Comment (metanerd) [inline server/latest.go]:
```suggestion
				_ = addRoutingRule(*value.Key, fileSearchValue, params, "-dmg")
```
Could we change it to this please?

File: server/latest.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "failed to upload file, index.html")
+	}
+	LogInfo("File uploaded to, %s\n", result.Location)
+
+	return nil
+}
+
+// SetLatestURL updates the S3 website routing configuration
+func SetLatestURL(typeToRelease string, ver string, cfg *MatterbuildConfig) error {
```
Comment (metanerd) [inline server/latest.go]:
```suggestion
func setLatestURL(typeToRelease string, ver string, cfg *MatterbuildConfig) error {
```

File: server/latest.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = preserverExistingRoutingRules(svc, cfg, typeToRelease, params)
+	if err != nil {
+		return err
+	}
+
+	generateNewRoutesForRelease(cfg, result, "mattermost-enterprise", ver, params)
```
Comment (metanerd) [inline server/latest.go]:
```suggestion
	err = generateNewRoutesForRelease(cfg, result, "mattermost-enterprise", ver, params)
```

File: server/latest.go (hunk 7 above)
Comment (metanerd) [reply server/latest.go]: Since this is the entry function, for readability purposes, I would put it on top of the file.

**Comment by metanerd** [comment]
My review seems like a lot, but it is really just nitpicking. Otherwise it looks really good! :tada:

**Comment by jaydeland** [comment]
> My review seems like a lot, but it is really just nitpicking. Otherwise it looks really good! tada

All good - thank you for the feedback - I believe the config changes in the private repo need to be merged first.

**Pull request merged**